File: crates/deepsec-cli/src/commands/revalidate.rs
use crate::backend::resolve_backend;
use crate::config_loader::Context;
use anyhow::Result;
use clap::Args as ClapArgs;
use colored::Colorize;
use deepsec_processor::{RevalidateOptions, run_revalidate};

#[derive(Debug, ClapArgs)]
pub struct Args {
    #[arg(long)]
    pub project_id: String,
    #[arg(long)]
    pub agent: Option<String>,
    #[arg(long)]
    pub model: Option<String>,
    #[arg(long)]
    pub filter: Option<String>,
    #[arg(long)]
    pub force: bool,
}

pub async fn run(args: Args, ctx: &Context) -> Result<()> {
    let proj = ctx.project(&args.project_id)?;
    let default_agent = ctx.config.as_ref().and_then(|c| c.default_agent.as_deref());
    crate::preflight::check(args.agent.as_deref().or(default_agent))?;
    let backend = resolve_backend(args.agent.as_deref(), args.model.as_deref(), default_agent)?;
    let opts = RevalidateOptions {
        project_id: args.project_id.clone(),
        project_root: proj.root,
        data_root: ctx.data_root.clone(),
        backend,
        filter_prefix: args.filter,
        force: args.force,
    };
    let out = run_revalidate(opts).await?;
    println!(
        "{} run={} revalidated={} TP={} FP={} fixed={} uncertain={}",
        "revalidate".bold().green(),
        out.run_id,
        out.revalidated,
        out.true_positives,
        out.false_positives,
        out.fixed,
        out.uncertain
    );
    Ok(())
}

File: internal/cli/commands/revalidate.go
package commands

import (
	"context"
	"fmt"

	"github.com/noeljackson/deepsec/internal/cli"
	"github.com/noeljackson/deepsec/internal/processor"
	"github.com/spf13/cobra"
)

// NewRevalidateCmd runs the revalidation stage.
func NewRevalidateCmd(loader func() (*cli.Context, error)) *cobra.Command {
	var projectID, agent, model, filter string
	var force bool
	var temperature, topP float64
	var seed int64
	cmd := &cobra.Command{
		Use:   "revalidate",
		Short: "Re-verify existing findings against the current code",
		RunE: func(cmd *cobra.Command, _ []string) error {
			ctx, err := loader()
			if err != nil {
				return err
			}
			proj, err := ctx.Project(projectID)
			if err != nil {
				return err
			}
			agentName := ctx.ResolveAgent(agent)
			if err := cli.Preflight(ctx, agentName); err != nil {
				return err
			}
			profile := ctx.Providers.Get(agentName)
			apiKey, err := ctx.Providers.LookupKey(agentName)
			if err != nil {
				return err
			}
			settings := modelSettingsFromFlags(cmd, temperature, topP, seed)
			backend, err := processor.NewBackend(profile, model, apiKey, settings)
			if err != nil {
				return err
			}
			out, err := processor.Revalidate(context.Background(), processor.RevalidateOptions{
				ProjectID:     projectID,
				ProjectRoot:   proj.Root,
				DataRoot:      ctx.DataRoot,
				Backend:       backend,
				ProviderName:  agentName,
				FilterPrefix:  filter,
				Force:         force,
				ModelSettings: settings,
			})
			if err != nil {
				return err
			}
			fmt.Printf("revalidate run=%s revalidated=%d TP=%d FP=%d fixed=%d uncertain=%d\n",
				out.RunID, out.Revalidated, out.TruePositives, out.FalsePositives, out.Fixed, out.Uncertain)
			return nil
		},
	}
	cmd.Flags().StringVar(&projectID, "project-id", "", "Project id (required)")
	_ = cmd.MarkFlagRequired("project-id")
	cmd.Flags().StringVar(&agent, "agent", "", "Provider name")
	cmd.Flags().StringVar(&model, "model", "", "Override the backend model")
	cmd.Flags().StringVar(&filter, "filter", "", "Path-prefix filter")
	cmd.Flags().BoolVar(&force, "force", false, "Re-revalidate findings that already have a verdict")
	cmd.Flags().Float64Var(&temperature, "temperature", 0, "Pin sampling temperature (default: provider SDK default)")
	cmd.Flags().Float64Var(&topP, "top-p", 0, "Pin nucleus sampling (default: provider SDK default)")
	cmd.Flags().Int64Var(&seed, "seed", 0, "Pin sampler seed; OpenAI-compatible providers only")
	return cmd
}
